Rodeos http wasm ql timeout 📦

shahabi8 opened the pull request:
<!-- PLEASE FILL OUT THE FOLLOWING MARKDOWN TEMPLATE -->
<!-- PR title alone should be sufficient to understand changes. -->
This PR adds timeout to http connection to rodeos. A python test is also added to verify the functionality of the timer for both TCP and unix-socket connections. 


## Change Type
**Select *ONE*:**
- [ ] Documentation
<!-- checked [x] = Documentation; unchecked [ ] = no changes, ignore this section -->
- [X] Stability bug fix
<!-- checked [x] = Stability bug fix; unchecked [ ] = no changes, ignore this section -->
- [ ] Other
<!-- checked [x] = Other; unchecked [ ] = no changes, ignore this section -->
- [ ] Other - special case
<!-- checked [x] = Other - special case; unchecked [ ] = no changes, ignore this section -->
<!-- Other - special case is for when a change warrants additional explanation or description in the release notes. Please include a description of the change for inclusion in the release notes. -->


## Testing Changes
**Select *ANY* that apply:**
- [X] New Tests
<!-- checked [x] = new test cases were added; unchecked [ ] = no new test cases -->
- [ ] Existing Tests
<!-- checked [x] = existing test cases were edited; unchecked [ ] = no existing tests were modified -->
- [ ] Test Framework
<!-- checked [x] = this modifies the test framework; unchecked [ ] = no test framework changes -->
- [ ] CI System
<!-- checked [x] = this changes the CI system; unchecked [ ] = no CI changes -->
- [ ] Other
<!-- checked [x] = this integrates an external test system; unchecked [ ] = no miscellaneous test-related changes -->
<!-- Please describe your test changes, or list each new test and its purpose, under each respective checkbox -->


## Consensus Changes
- [ ] Consensus Changes
<!-- checked [x] = Consensus changes; unchecked [ ] = no changes, ignore this section -->
<!-- If this PR introduces a change to the validation of blocks in the chain or consensus in general, please describe the impact. -->


## API Changes
- [ ] API Changes
<!-- checked [x] = API changes; unchecked [ ] = no changes, ignore this section -->
<!-- If this PR introduces API changes, please describe the changes here. What will developers need to know before upgrading to this version? -->


## Documentation Additions
- [ ] Documentation Additions
<!-- checked [x] = Documentation changes; unchecked [ ] = no changes, ignore this section -->
<!-- Describe what must be added to the documentation after merge. -->

shahabi8 committed:
Update generate-pipeline.sh

Added test name to exclude eos-vm-oc from MacOS

shahabi8 committed:
Update amazon_linux-2-pinned.dockerfile

shahabi8 committed:
Update centos-7.7-pinned.dockerfile

shahabi8 committed:
Update centos-8-pinned.dockerfile

shahabi8 committed:
Update macos-10.14-pinned.sh

shahabi8 committed:
Update macos-10.15-pinned.sh

shahabi8 committed:
Update ubuntu-16.04-pinned.dockerfile

shahabi8 committed:
Update ubuntu-18.04-pinned.dockerfile

shahabi8 committed:
Update ubuntu-20.04-pinned.dockerfile

shahabi8 committed:
Update amazon_linux-2-unpinned.dockerfile

shahabi8 committed:
Update centos-7.7-unpinned.dockerfile

shahabi8 committed:
Update centos-8-unpinned.dockerfile

shahabi8 committed:
Update macos-10.14-unpinned.sh

shahabi8 committed:
Update macos-10.15-unpinned.sh

shahabi8 committed:
Update ubuntu-18.04-unpinned.dockerfile

shahabi8 committed:
Update ubuntu-20.04-unpinned.dockerfile

shahabi8 committed:
Update wasm_ql_http.cpp

shahabi8 committed:
Update wasm_ql_http.hpp

shahabi8 committed:
Update wasm_ql_plugin.cpp

shahabi8 committed:
Update CMakeLists.txt

shahabi8 committed:
Update rodeos_utils.py

shahabi8 committed:
Create rodeos_wasmQL_http_timeout.py

linhuang-blockone commented on programs/rodeos/wasm_ql_plugin.cpp:
```diff
@@ -86,7 +86,7 @@ void wasm_ql_plugin::set_program_options(options_description& cli, options_descr
    op("wql-max-request-size", bpo::value<uint32_t>()->default_value(10000), "HTTP maximum request body size (bytes)");
-   op("wql-idle-timeout", bpo::value<uint64_t>()->default_value(std::numeric_limits<uint32_t>::max()),
+   op("wql-idle-timeout", bpo::value<uint32_t>()->default_value(30000),
       "HTTP idle connection timeout (currently no timeout supported, this option is ignored)");
```
This "(currently no timeout supported, this option is ignored)" should be removed.

shahabi8 replied:
Done.

linhuang-blockone commented on programs/rodeos/wasm_ql_http.cpp:
```diff
@@ -459,6 +461,10 @@ class http_session {
 
+      // Set the timeout.
+      session_begin = steady_clock::now();
+      //derived_session().stream.expires_after(std::chrono::milliseconds(http_config->idle_timeout_ms));
```
Remove the comment.

shahabi8 replied:
Done.

linhuang-blockone commented on programs/rodeos/wasm_ql_http.cpp:
```diff
@@ -470,7 +476,11 @@ class http_session {
+      auto session_time = steady_clock::now() - session_begin;
+      auto session_time_ms = std::chrono::duration_cast<std::chrono::milliseconds>( session_time ).count();
+      elog("${session_time_ms}: ${idle_timeout_ms}", ("session_time_ms", session_time_ms)("idle_timeout_ms", http_config->idle_timeout_ms.count()));
+      if (session_time_ms > http_config->idle_timeout_ms.count()){
```
should be if (ec || session_time_ms > http_config->idle_timeout_ms.count())

shahabi8 replied:
In here I set error code (ec) for timeout, so I added another if statement for other error codes.

linhuang-blockone commented on programs/rodeos/wasm_ql_http.cpp:
```diff
@@ -470,7 +476,11 @@ class http_session {
-      if (ec) {
+      auto session_time = steady_clock::now() - session_begin;
+      auto session_time_ms = std::chrono::duration_cast<std::chrono::milliseconds>( session_time ).count();
+      elog("${session_time_ms}: ${idle_timeout_ms}", ("session_time_ms", session_time_ms)("idle_timeout_ms", http_config->idle_timeout_ms.count()));
```
why elog before the `if` check?

shahabi8 replied:
I removed this elog, it was only for debugging purposes.

linhuang-blockone commented on .cicd/platforms/pinned/amazon_linux-2-pinned.dockerfile:
```diff
@@ -13,8 +13,9 @@ RUN curl -s https://packagecloud.io/install/repositories/rabbitmq/erlang/script.
-RUN pip3 install --upgrade pip
+# upgrade pip installation. request and requests_unixsocket module
+RUN pip3 install --upgrade pip && \
+    pip3 install requests && pip3 install requests_unixsocket
```
A general comment: I am not sure it is worthy the trouble to install request and requests_unixsocket in all those scripts.

shahabi8 replied:
In develop branch we have `requests` module in base images, but we don't have it here in develop-box, so I added it here too and also `request_unixsocket`. If there are other methods to test persistence connection, please let me know.

shahabi8 committed:
Update wasm_ql_http.cpp

Removed unused comment, elog, and added if-statement for all other error codes

shahabi8 committed:
Update wasm_ql_plugin.cpp

Removed comment.

heifner reviewed (changes requested):
Maybe I missed it, but how does this actually time out a connection. Looks like it will timeout a connection on the next request, but if no request is made how does it timeout and close the connection?

shahabi8 committed:
Update wasm_ql_http.cpp

Added asio timer to wasmQL

shahabi8 committed:
Update wasm_ql_plugin.cpp

Set timer default value

shahabi8 committed:
Update CMakeLists.txt

Modified the added test

shahabi8 committed:
Update rodeos_wasmQL_http_timeout.py

New version of test

shahabi8 committed:
Update generate-pipeline.sh

Removed the eos-vm tests from running on mac

shahabi8 committed:
Update wasm_ql_http.cpp

Fixed a syntax error

shahabi8 committed:
Merge branch 'rodeos_tests_develop_boxed' into rodeos_http_wasmQL_timeout

shahabi8 committed:
Update wasm_ql_http.cpp

shahabi8 committed:
Update wasm_ql_plugin.cpp

linhuang-blockone commented on programs/rodeos/wasm_ql_http.cpp:
```diff
@@ -445,12 +447,37 @@ class http_session {
+      _timer.reset(new boost::asio::steady_timer(derived_session().stream.socket().get_executor()));
+      session_begin = steady_clock::now();
+      _timer->expires_after( http_config->idle_timeout_ms );
+      _timer->async_wait(boost::asio::bind_executor(derived_session().stream.socket().get_executor(), boost::bind(&http_session::wait_to_timeout, derived_session().shared_from_this())));
```
I don't see where this timer is cancelled? Shouldn't it be cancelled when a message comes in?

shahabi8 replied:
Timer gets cancelled through `wait_to_timeout` method which is called inside `boost::bind::executor()`

linhuang-blockone commented on programs/rodeos/wasm_ql_plugin.cpp:
```diff
@@ -86,7 +86,7 @@ void wasm_ql_plugin::set_program_options(options_description& cli, options_descr
    op("wql-wasm-cache-size", bpo::value<uint32_t>()->default_value(100), "Maximum number of compiled wasms to cache");
    op("wql-max-request-size", bpo::value<uint32_t>()->default_value(10000), "HTTP maximum request body size (bytes)");
-   op("wql-idle-timeout", bpo::value<uint32_t>()->default_value(30000), "HTTP idle connection timeout");
+   op("wql-idle-timeout", bpo::value<uint64_t>()->default_value(std::numeric_limits<uint32_t>::max()), "HTTP idle connection timeout");
```
add (ms) after timeout so users know what time unit is used.

shahabi8 replied:
Done.

shahabi8 committed:
Update wasm_ql_plugin.cpp

shahabi8 committed:
Update wasm_ql_http.cpp

Changing timer reset.

shahabi8 committed:
Update wasm_ql_http.cpp

shahabi8 committed:
Update rodeos_utils.py

shahabi8 committed:
Update rodeos_utils.py

shahabi8 committed:
Update rodeos_utils.py

shahabi8 committed:
Update wasm_ql_http.cpp

Made some changes such that timer gets cancelled along with socket inside do_close()

shahabi8 committed:
Update wasm_ql_http.cpp

shahabi8 committed:
Merge branch 'rodeos_tests_develop_boxed' into rodeos_http_wasmQL_timeout

heifner commented on programs/rodeos/wasm_ql_http.cpp:
```diff
@@ -489,12 +492,37 @@ class http_session {
+         }
+         auto session_duration    = steady_clock::now() - last_activity_timepoint;
+         auto session_duration_ms = std::chrono::duration_cast<std::chrono::milliseconds>( session_duration ).count();
+         if ( session_duration_ms <= http_config->idle_timeout_ms.count() ){
```
This can be simplified to:

```
         auto session_duration = steady_clock::now() - last_activity_timepoint;
         if ( session_duration <= http_config->idle_timeout_ms ){
```

shahabi8 replied:
Thanks for catching this, fixed it.

shahabi8 committed:
Update wasm_ql_http.cpp

Added requested changes + removing declaration of beast::error
